src/network/core/packet.cpp

```diff
@@ -264,6 +291,15 @@ void Packet::PrepareToRead()
 {
 	/* Put the position on the right place */
 	this->pos = sizeof(PacketSize);
+
+	if (cs == nullptr || cs->receive_encryption_handler == nullptr) return;
+
+	size_t mac_size = cs->receive_encryption_handler->MACSize();
+	if (this->buffer.size() >= 1 + pos + mac_size) {
+		/* There is at least enough data to have something to decompress. */
+		cs->receive_encryption_handler->Decrypt(std::span(&this->buffer[pos], mac_size), std::span(&this->buffer[pos + mac_size], this->buffer.size() - pos - mac_size));
```

Comment:
Does this means that *any* small packet can be injected into connection and it will be accepted? Doesn't seem right.

Reply:
It also looks like packet type field isn't protected.

Reply:
> Does this means that _any_ small packet can be injected into connection and it will be accepted? Doesn't seem right.

Yes it does. Though I'm running into another issue while implementing authorized keys, so this needs (way) more thought. The problem is that authentication of authorized key can fail, and then it should retry. However, the connection is already encrypted at the client by that point while the server has not enabled encryption yet because the authentication has not succeeded yet.


> It also looks like packet type field isn't protected.

What do you mean? The packet type is the first byte after the MAC, so it should be included in the encrypted. Or is this about `protected` in the sense of access specifiers?

Reply:
> Yes it does.

Perhaps it could enforce that packet is of ERROR type in that case? Though the implicit assumptions that errors are small feels weird.

> The packet type is the first byte after the MAC, so it should be included in the encrypted.

Ah sorry, I didn't notice that.